File: 2019W07/preppindataw7.R
library(readxl)
library(dplyr)

deptdtl <- read_excel("E:/Week7Challenge.xlsx", 
                           sheet = "Departure Details", col_types = c("text", 
                                                                      "date", "numeric", "numeric"))

allcdtl <- read_excel("E:/Week7Challenge.xlsx", 
                            sheet = "Allocation Details", col_types = c("text", 
                                                                        "text", "date", "text", "numeric", 
                                                                        "numeric"))

final <- deptdtl %>%
      mutate("Departure ID" = paste0(`Ship ID`,"-",format(`Departure Date`, format="%d-%m-%Y"))) %>%
      merge(., allcdtl, by="Departure ID") %>%
      group_by(`Ship ID`,`Departure Date`) %>% 
      summarise("Max Weight"= max(`Max Weight`),
                "Max Volume"= max(`Max Volume`),
                "Weight Allocated"= sum(`Weight Allocated`),
                "Volume Allocated"= sum(`Volume Allocated`)) %>%
      mutate("Weight Exceeded" = `Weight Allocated`>`Max Weight`,
             "Volume Exceeded" = `Volume Allocated`>`Max Volume`,
             "Departure Date" = format(`Departure Date`,format="%d/%m/%Y")) 


View(final)

File: 2019/2019W07/PreppindataW7.py
import pandas as pd

xls = pd.ExcelFile("E:\Week7Challenge.xlsx")

dept = pd.read_excel(xls,"Departure Details") 
allc = pd.read_excel(xls,"Allocation Details") 

dept["Departure ID"]=dept["Ship ID"].map(str)+dept["Departure Date"].apply(lambda x: x.strftime('-%d-%m-%Y'))
dept["Departure Date"] = dept["Departure Date"].dt.strftime('%d/%m/%Y')

temp = dept.merge(allc, on="Departure ID", how='inner')

aggr = {
    'Max Weight':'max',
    'Max Volume':'max',
    'Weight Allocated':'sum',
    'Volume Allocated':'sum'
}

final = temp.groupby(["Ship ID","Departure Date"], as_index=False).agg(aggr)
final["Weight Exceeded"] = final.apply(lambda x: x['Weight Allocated']>x['Max Weight'],axis=1)
final["Volume Exceeded"] = final.apply(lambda x: x['Volume Allocated']>x['Max Volume'],axis=1)


